sdk/communication/azure-communication-chat/src/main/java/com/azure/android/communication/chat/implementation/notifications/fcm/RegistrationKeyManager.java

```diff
@@ -64,84 +94,266 @@ public boolean getLastExecutionSucceeded() {
+        try (FileInputStream fis2 = new File(keyCreationTimeStorePath).exists() ? new FileInputStream(keyCreationTimeStorePath) : null) {
+            keyCreationTimeStore.load(fis2);
         } catch (IOException e) {
-            clientLogger.logExceptionAsError(new RuntimeException("Failed to load key store", e));
+            throw clientLogger.logExceptionAsError(new RuntimeException("Failed to load key creation time store", e));
         }
     }
 
-    private String secretKeyToStr(SecretKey secretKey) {
-        return Base64Util.encodeToString(secretKey.getEncoded());
+    private char[] getPassword() {
+        return "com.azure.android.communication.chat.android".toCharArray();
     }
 
-    //The last pair of secrete keys are to be used to renew registration
-    public RegistrationKeyStore.RegistrationKeyEntry getLastEntry() {
-        synchronized (RegistrationKeyManager.class) {
-            return registrationKeyStore.getLastEntry();
+    private void deleteKeyFromFiles(String alias, String directoryPath) {
+        //delete from key-store
+        String keyStorePath = directoryPath + KEY_STORE_POSTFIX;
+        try {
+            keyStore.deleteEntry(alias);
+        } catch (KeyStoreException e) {
+            throw clientLogger.logExceptionAsError(new RuntimeException("Failed to delete entry from key-store with alias: " + alias, e));
+        }
+
+        File outputFile = new File(keyStorePath);
+        if (!outputFile.exists()) {
+            try {
+                outputFile.createNewFile();
+            } catch (IOException e) {
+                throw clientLogger.logExceptionAsError(new RuntimeException("Failed to create key store file", e));
+            }
+        }
+        try (FileOutputStream fos = new FileOutputStream(outputFile)) {
+            keyStore.store(fos, getPassword());
+        } catch (Exception e) {
+            throw clientLogger.logExceptionAsError(new RuntimeException("Failed to save secrete key", e));
+        }
+
+        //delete from key-creation-time-store
+        String keyCreationTimeStorePath = directoryPath + KEY_CREATION_TIME_POSTFIX;
+        try {
+            keyCreationTimeStore.deleteEntry(keyCreationTimeStorePath, alias);
+        } catch (Exception e) {
+            RuntimeException runtimeException = new RuntimeException("keyCreationTimeStore failed to delete entry");
+            throw clientLogger.logExceptionAsError(runtimeException);
+        }
+    }
+
+    public Long getCreationTime(String alias) {
+        return keyCreationTimeStore.getCreationTime(alias);
+    }
+
+    public SecretKey getSecreteKey(String alias) {
+        KeyStore.ProtectionParameter protParam =
+            new KeyStore.PasswordProtection(getPassword());
+
+        // get my private key
+        KeyStore.SecretKeyEntry pkEntry = null;
+        try {
+            pkEntry = (KeyStore.SecretKeyEntry) keyStore.getEntry(alias, protParam);
```

Comment:
I don't think this will play nice if the result of `getEntry()` is null. I think we should check for null before attempting casting to `KeyStore.SecretKeyEntry`.

Reply:
There is a null checking before converting entry to secretKey.